pkg/profile/artifact.go

```diff
@@ -7,54 +7,180 @@ import (
 	"time"
 
 	kustomizev1 "github.com/fluxcd/kustomize-controller/api/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 
 	helmv2 "github.com/fluxcd/helm-controller/api/v2beta1"
 	sourcev1 "github.com/fluxcd/source-controller/api/v1beta1"
 )
 
+type Status struct {
+	ResourcesExist     bool
+	NotReadyConditions []metav1.Condition
+}
+
 // CreateArtifacts creates and inserts objects to the cluster to deploy the
 // profile as a HelmRelease.
 func (p *Profile) CreateArtifacts(ctx context.Context) error {
-	if err := p.createGitRepository(ctx); err != nil {
+	gitRes, helmRes, kustomizeRes, err := p.makeArtifacts()
+	if err != nil {
+		return err
+	}
+
+	p.log.Info("creating GitRepository", "resource", gitRes.ObjectMeta.Name)
+	if err := p.client.Create(ctx, gitRes); err != nil {
 		return fmt.Errorf("failed to create GitRepository resource: %w", err)
 	}
 
-	switch kind := p.definition.Spec.Artifacts[0].Kind; kind {
-	case "HelmChart":
-		if err := p.createHelmRelease(ctx); err != nil {
+	if helmRes != nil {
+		p.log.Info("creating HelmRelease", "resource", helmRes.ObjectMeta.Name)
+		if err := p.client.Create(ctx, helmRes); err != nil {
 			return fmt.Errorf("failed to create HelmRelease resource: %w", err)
 		}
-	case "Kustomize":
-		if err := p.createKustomization(ctx); err != nil {
+	} else {
+		p.log.Info("creating Kustomization", "resource", kustomizeRes.ObjectMeta.Name)
+		if err := p.client.Create(ctx, kustomizeRes); err != nil {
 			return fmt.Errorf("failed to create Kustomization resource: %w", err)
 		}
-	default:
-		return fmt.Errorf("artifact kind %q not recognized", kind)
 	}
-
 	p.log.Info("all artifacts created")
 	return nil
 }
 
+// Checks if the artifacts exists and returns any ready!=true conditions on the artifacts.
+func (p *Profile) ArtifactStatus(ctx context.Context) (Status, error) {
+	resourcesExist, gitRes, helmRes, kustomizeRes, err := p.getResources(ctx)
+	if err != nil {
+		return Status{}, err
+	}
+
+	return Status{
+		ResourcesExist:     resourcesExist,
+		NotReadyConditions: p.checkResourcesReady(gitRes, helmRes, kustomizeRes),
+	}, nil
+}
+
+func (p *Profile) checkResourcesReady(gitRes *sourcev1.GitRepository, helmRes *helmv2.HelmRelease, kustomizeRes *kustomizev1.Kustomization) []metav1.Condition {
+	var notReadyConditions []metav1.Condition
+	if gitRes != nil {
+		if condition := getNotReadyCondition(gitRes.Status.Conditions); condition != (metav1.Condition{}) {
```

Comment:
The change to getNotReady would then result in this:

```go
		if ok, condition := getNotReadyCondition(gitRes.Status.Conditions); ok {
			notReadyConditions = append(notReadyConditions, condition)
		}
```
which reads nice, I think :)